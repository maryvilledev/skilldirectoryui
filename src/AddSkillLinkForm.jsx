import React from 'react';
import axios from 'axios';
import 'react-select/dist/react-select.css';
import 'bootstrap/dist/css/bootstrap.css';
import { Col, ControlLabel, Button,
   Form, FormControl, FormGroup } from 'react-bootstrap'

class LinksForm extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      link_name: "",
      link_url: "",
      link_type: "",
    }
    this.onChange = this.onChange.bind(this);
    this.onSubmit = this.onSubmit.bind(this);
  }

  onChange(key, value) {
    this.setState({ [key]: value });
  }

  onSubmit(ev) {
    ev.preventDefault();
    // Error checking
    if(this.state.link_name === '') {
      alert('Please enter a Name.');
      return;
    }
    if(!isValidURL(this.state.link_url)) {
      alert('Please enter a valid URL.');
      return;
    }
    if(this.state.link_type === '') {
      alert('Please enter a Type.');
      return;
    }

    // Post form data to API endpoint
    axios.post(this.props.api + '/links/', {
      name:      this.state.link_name,
      url:       this.state.link_url,
      skill_id:  this.props.skill_id,
      link_type: this.state.link_type
    })
    .then(response => {
      this.props.closeModal();
    }).catch(err => console.log(`Caught an error ${err}`));
  }

  render() {
    const onLinkNameChange = ev => this.onChange('link_name', ev.target.value);
    const onLinkURLChange = ev => this.onChange('link_url', ev.target.value);
    const onLinkTypeChange = ev => this.onChange('link_type', ev.target.value);

    return (
      <div>
        <h1>Add a Link</h1>
        <Form horizontal onSubmit={this.onSubmit}>
          <FormGroup controlId='linkName'>
            <Col componentClass={ControlLabel} sm={2}>
              Name:
            </Col>
            <Col sm={10}>
              <FormControl
                name='linkName'
                componentClass='input'
                onChange={onLinkNameChange}/>
            </Col>
          </FormGroup>
          
          <FormGroup controlId='linkURL'>
            <Col componentClass={ControlLabel} sm={2}>
              URL:
            </Col>
            <Col sm={10}>
              <FormControl
                name='linkURL'
                componentClass='input'
                onChange={onLinkURLChange}/>
            </Col>
          </FormGroup>

          <FormGroup controlId='linkType'>
            <Col componentClass={ControlLabel} sm={2}>
              Type:
            </Col>
            <Col sm={10}>
              <FormControl
                name='linkType'
                componentClass='select'
                onChange={onLinkTypeChange}>
                <option key={null} value={null}/>
                <option key='blog' value='blog'>
                  Blog
                </option>
                <option key='tutorial' value='tutorial'>
                  Tutorial
                </option>
                <option key='webpage' value='webpage'>
                  Webpage
                </option>
              </FormControl>
            </Col>  
          </FormGroup>

          <FormGroup>
            <Col smOffset={2} sm={2}>
              <Button type='submit' bsStyle='primary'>
                Submit 
              </Button>
            </Col>
            <Col smOffset={2}>
              <Button  onClick={this.props.closeModal} bsStyle='info'>
                Cancel 
              </Button>
            </Col>
          </FormGroup>
        </Form>
      </div>
    );
  }
}

/* eslint-disable no-useless-escape */
/* http://stackoverflow.com/a/30970319 */
function isValidURL(url) {
    var res = url.match(/(http(s)?:\/\/.)(www\.)?[-a-zA-Z0-9@:%._\+~#=]{2,256}\.[a-z]{2,6}\b([-a-zA-Z0-9@:%_\+.~#?&//=]*)/g);
    if(res == null)
        return false;
    else
        return true;
}

export default LinksForm
module.exports.isValidURL = isValidURL
