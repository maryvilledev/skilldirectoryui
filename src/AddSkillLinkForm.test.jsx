import React from 'react';
import ReactDOM from 'react-dom';
import { shallow, mount } from 'enzyme';

import AddSkillLinkForm, { isValidURL } from './AddSkillLinkForm.jsx';

describe('<AddSkillLinkForm />', () => {
  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<AddSkillLinkForm />, div);
  });

  it('matches the stored snapshot', () => {
    const wrapper = shallow(<AddSkillLinkForm />);
    expect(wrapper.get(0)).toMatchSnapshot();
  });

  it('changes state when input to "Name" field changes', () => {
    const wrapper = mount(<AddSkillLinkForm />);
    expect(wrapper.state().linkName).toBe('');
    const event = { target: { value: 'New Value' } };
    wrapper.find({ name: 'linkName' }).simulate('change', event);
    expect(wrapper.state().linkName).toBe('New Value');
  });

  it('changes state when input to "URL" field changes', () => {
    const wrapper = mount(<AddSkillLinkForm />);
    expect(wrapper.state().linkUrl).toBe('');
    const event = { target: { value: 'New Value' } };
    wrapper.find({ name: 'linkURL' }).simulate('change', event);
    expect(wrapper.state().linkUrl).toBe('New Value');
  });

  it('changes state when selected option in "Type" field changes', () => {
    const wrapper = mount(<AddSkillLinkForm />);
    expect(wrapper.state().linkType).toBe('');
    const newType = 'tutorial';
    const event = { target: { value: newType } };
    wrapper.find({ name: 'linkType' }).simulate('change', event);
    expect(wrapper.state().linkType).toBe(newType);
  });
});

test('isValidURL() successfully detects valid/invalid URL forms', () => {
  expect(isValidURL('http://www.google.com')).toBe(true);
  expect(isValidURL('https://www.github.com')).toBe(true);
  expect(isValidURL('www.google.com')).toBe(false);
  expect(isValidURL('definitely-not-a-valid-url')).toBe(false);
});
