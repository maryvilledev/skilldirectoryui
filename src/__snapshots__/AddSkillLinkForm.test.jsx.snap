exports[`<AddSkillLinkForm /> matches the stored snapshot 1`] = `
<div>
  <h1>
    Add a Link
  </h1>
  <Form
    bsClass="form"
    componentClass="form"
    horizontal={true}
    inline={false}
    onSubmit={[Function]}>
    <FormGroup
      bsClass="form-group"
      controlId="linkName">
      <Col
        bsClass="col"
        componentClass={[Function]}
        sm={2}>
        Name:
      </Col>
      <Col
        bsClass="col"
        componentClass="div"
        sm={10}>
        <FormControl
          bsClass="form-control"
          componentClass="input"
          name="linkName"
          onChange={[Function]} />
      </Col>
    </FormGroup>
    <FormGroup
      bsClass="form-group"
      controlId="linkURL">
      <Col
        bsClass="col"
        componentClass={[Function]}
        sm={2}>
        URL:
      </Col>
      <Col
        bsClass="col"
        componentClass="div"
        sm={10}>
        <FormControl
          bsClass="form-control"
          componentClass="input"
          name="linkURL"
          onChange={[Function]} />
      </Col>
    </FormGroup>
    <FormGroup
      bsClass="form-group"
      controlId="linkType">
      <Col
        bsClass="col"
        componentClass={[Function]}
        sm={2}>
        Type:
      </Col>
      <Col
        bsClass="col"
        componentClass="div"
        sm={10}>
        <FormControl
          bsClass="form-control"
          componentClass="select"
          name="linkType"
          onChange={[Function]}>
          <option
            value={null} />
          <option
            value="blog">
            Blog
          </option>
          <option
            value="tutorial">
            Tutorial
          </option>
          <option
            value="webpage">
            Webpage
          </option>
        </FormControl>
      </Col>
    </FormGroup>
    <FormGroup
      bsClass="form-group">
      <Col
        bsClass="col"
        componentClass="div"
        sm={2}
        smOffset={2}>
        <Button
          active={false}
          block={false}
          bsClass="btn"
          bsStyle="primary"
          disabled={false}
          type="submit">
          Submit
        </Button>
      </Col>
      <Col
        bsClass="col"
        componentClass="div"
        smOffset={2}>
        <Button
          active={false}
          block={false}
          bsClass="btn"
          bsStyle="info"
          disabled={false}
          onClick={undefined}>
          Cancel
        </Button>
      </Col>
    </FormGroup>
  </Form>
</div>
`;
